src/box/box.cc

```diff
@@ -1805,6 +1811,230 @@ box_wait_quorum(uint32_t lead_id, int64_t target_lsn, int quorum,
+ * When an ack comes with requested sync for some replica id, ack vclock is
+ * accounted.
+ */
+static int
+check_vclock_sync_on_ack(struct trigger *trigger, void *event)
+{
+	struct replication_ack *ack = (struct replication_ack *)event;
+	struct sync_trigger_data *data =
+		(struct sync_trigger_data *)trigger->data;
+	uint32_t id = ack->source;
+	uint64_t sync = data->vclock_syncs[id];
+	int accounted_count = bit_count_u32(data->collected_vclock_map) + 1;
+	if (!bit_test(&data->collected_vclock_map, id) && sync > 0 &&
+	    ack->vclock_sync >= sync && accounted_count < data->count) {
+		vclock_max_ignore0(data->vclock, ack->vclock);
+		bit_set(&data->collected_vclock_map, id);
+		++accounted_count;
+		if (accounted_count >= data->count)
+			fiber_wakeup(data->waiter);
+	}
+	return 0;
+}
+
+/** A trigger querying relay's next sync value once it becomes operational. */
+static int
+relay_get_sync_on_start(struct trigger *trigger, void *event)
+{
+	struct replica *replica = (struct replica *)event;
+	if (replica->anon)
+		return 0;
+	struct relay *relay = replica->relay;
+	struct sync_trigger_data *data =
+		(struct sync_trigger_data *)trigger->data;
+	uint32_t id = replica->id;
+	/* Already accounted. */
+	if (bit_test(&data->collected_vclock_map, id))
+		return 0;
+	if (relay_next_vclock_sync(relay, &data->vclock_syncs[id],
+				   data->deadline) != 0) {
+		diag_clear(diag_get());
+		data->is_timed_out = true;
+		fiber_wakeup(data->waiter);
+	}
+	return 0;
+}
+
+/** Find the minimal vclock which has all the data confirmed on a quorum. */
+static int
+box_collect_confirmed_vclock(struct vclock *confirmed_vclock, double deadline)
+{
+	/*
+	 * How many vclocks we should see to be sure that at least one of them
+	 * contains all data present on any real quorum.
+	 */
+	int vclock_count = MAX(1, replicaset.registered_count) -
+			   replication_synchro_quorum + 1;
+	/*
+	 * We should check the vclock on self plus vclock_count - 1 remote
+	 * instances.
+	 */
+	vclock_copy(confirmed_vclock, &replicaset.vclock);
+	if (vclock_count <= 1)
+		return 0;
+
+	struct sync_trigger_data data = {
+		.vclock_syncs = {0},
+		.collected_vclock_map = 0,
+		.waiter = fiber(),
+		.vclock = confirmed_vclock,
+		.deadline = deadline,
+		.count = vclock_count,
+		.is_timed_out = false,
+	};
+	struct trigger on_relay_thread_start;
+	trigger_create(&on_relay_thread_start, relay_get_sync_on_start, &data,
+		       NULL);
+	trigger_add(&replicaset.on_relay_thread_start, &on_relay_thread_start);
+	struct trigger on_ack;
+	trigger_create(&on_ack, check_vclock_sync_on_ack, &data, NULL);
+	trigger_add(&replicaset.on_ack, &on_ack);
+	replicaset_foreach(replica) {
+		if (relay_get_state(replica->relay) != RELAY_FOLLOW ||
+		    replica->anon) {
+			continue;
+		}
+		/* Might be already filled by on_relay_thread_start trigger. */
+		if (data.vclock_syncs[replica->id] != 0)
+			continue;
+		if (relay_next_vclock_sync(replica->relay,
```

Comment:
Better call `relay_trigger_vclock_sync` or even simply `relay_sync_vclock` IMO.

Reply:
Yep, `relay_trigger_vclock_sync` makes more sense. Done.
`relay_sync_vclock` would look like the returned sync value is already reached IMO.